src/Development/Shake/Internal/Args.hs

```diff
@@ -229,43 +232,79 @@ shakeArgsOptionsWith baseOpts userOptions rules = do
-                                    actionShareSanity
-                            withoutActions rules
+                    let rules2 = if NoBuild `elem` flagsExtra then
+                                withoutActions rules
+                            else if ShareList `elem` flagsExtra ||
+                                    not (null shareRemoves) ||
+                                    ShareSanity `elem` flagsExtra then do
+                                action $ do
+                                    unless (null shareRemoves) $
+                                        actionShareRemove shareRemoves
+                                    when (ShareList `elem` flagsExtra)
+                                        actionShareList
+                                    when (ShareSanity `elem` flagsExtra)
+                                        actionShareSanity
+                                withoutActions rules
+                            else
+                                rules
+
+                    let maybeWatch | shakeWatch shakeOpts = watch
+                                   | otherwise            = shakeWithDatabase
+
+                    maybeWatch shakeOpts rules2 $ \db -> do
+                        res <- try_ $ do
+                            (_, after) <- shakeRunDatabase db []
+                            shakeRunAfter shakeOpts after
+
+                        if shakeVerbosity shakeOpts < Info || NoTime `elem` flagsExtra then
+                            either throwIO pure res
                         else
-                            rules
-                    pure (True, shakeOpts, res)
-
-        if not ran || shakeVerbosity shakeOpts < Info || NoTime `elem` flagsExtra then
-            either throwIO pure res
-         else
-            let esc = if shakeColor shakeOpts then escape else \_ x -> x
-            in case res of
-                Left err ->
-                    if Exception `elem` flagsExtra then
-                        throwIO err
-                    else do
-                        putWhenLn Error $ esc Red $ show err
-                        exitFailure
-                Right () -> do
-                    tot <- start
-                    putWhenLn Info $ esc Green $ "Build completed in " ++ showDuration tot
-
+                            let esc = if shakeColor shakeOpts then escape else \_ x -> x
+                            in case res of
+                                Left err ->
+                                    if Exception `elem` flagsExtra then
+                                        throwIO err
+                                    else do
+                                        putWhenLn Error $ esc Red $ show err
+                                        exitFailure
+                                Right () -> do
+                                    tot <- start
+                                    putWhenLn Info $ esc Green $ "Build completed in " ++ showDuration tot
```

Comment:
Turn off whitespace changes when reviewing this section - it's not as complex as it looks:

<img width="1927" alt="CleanShot 2023-03-23 at 16 36 38@2x" src="https://user-images.githubusercontent.com/1387653/227380163-a0c05621-8504-42bc-bc0b-d257deb1192d.png">